relay/cache/cache.go

```diff
@@ -0,0 +1,30 @@
+package cache
+
+// WeightCalculator is a function that calculates the weight of a key-value pair in a Cache.
+// By default, the weight of a key-value pair is 1. Cache capacity is always specified in terms of
+// the weight of the key-value pairs it can hold, rather than the number of key-value pairs.
+//
+// Unless otherwise noted, Cache implementations are not required to be thread safe.
```

Comment:
why? It's quite often the use cases will need thread safety.

Also please move this line down to the Cache

Reply:
With the cache accessor, we are forced to use a mutex in order to perform atomic operations on the cache and the map of in-progress operations. Since that is currently our only use case, adding thread safety to the cache would be redundant and would provide no immediate benefit.

I agree that having a thread safe cache may be useful if we want to reuse it as a general purpose cache. I see two options:

- we make the cache inherently thread safe, even though not all use cases may need thread safety.
- we create a wrapper object that implements `Cache` that adds thread safety to another cache implementation, e.g. `func NewThreadSafeCache(cache Cache) Cache`. We could either write this now, or implement it later when we need it.

Of these options, which would you prefer?

Reply:
Currently there are no intended use cases outside of the cache accessor. Let's wait to add locking until we have a use case that justifies it.